vllm/v1/worker/gpu_model_runner.py

```diff
@@ -697,6 +701,44 @@ def __init__(
                                             pin_memory=pin_memory)
         self.top_k_cpu = self.top_k_cpu_tensor.numpy()
         self.top_k_reqs: Set[str] = set()
+        # Frequency penalty related data structures
+        self.frequency_penalties = torch.empty((max_num_reqs, vocab_size),
+                                               dtype=torch.float,
+                                               device=device)
+        self.frequency_penalties_cpu_tensor = torch.empty(
+            (max_num_reqs, vocab_size),
```

Comment:
Dumb question: Why is it `(max_num_reqs, vocab_size)` instead of just `(max_num_reqs, 1)`? Can we keep this tensor small and use broadcasting for actual computation?

Reply:
Done